### PR TITLE
Add remove_ids() function definition

### PR DESCRIPTION
**Pull request by jan-kozinski**
This adds definition for `Faiss::IndexIdMap2` and some missing Index methods

### CODE REVIEW

File: ext/faiss/index.cpp [hunk 1]
```diff
@@ -173,11 +174,45 @@ void init_index(Rice::Module& m) {
...
+        auto n = static_cast<std::size_t>(ni);
+        auto recons = numo::SFloat({n, d});
+        self.reconstruct_n(i0, ni, recons.write_ptr());
+        return recons;
+      })
+    .define_method(
+      "reconstruct_batch",
+      [](faiss::Index &self, numo::Int64 keys) {
+        if (keys.ndim() != 1) {
+          throw Rice::Exception(rb_eArgError, "expected keys to be 1d array");
```
Comment (ankane) [inline ext/faiss/index.cpp]: Let's use `ids` instead of `keys` in the error message.

File: test/index_test.rb [hunk 2]
```diff
@@ -319,6 +349,98 @@ def test_reconstruct
...
+  def test_remove_ids
+    objects = [
+      [1, 1, 2, 1],
+      [5, 4, 6, 5],
+      [1, 2, 1, 2]
+    ]
+    index = Faiss::IndexFlatL2.new(4)
+    index.add(objects)
+    index.remove_ids([0, 2])
+    assert_equal 1, index.ntotal
```
Comment (ankane) [inline test/index_test.rb:444]: This should test the id of the remaining record.

File: test/index_test.rb [hunk 3]
```diff
@@ -319,6 +349,98 @@ def test_reconstruct
...
+
+    # Reconstruct all vectors
+    recons = index.reconstruct_n(0, 5)
+    assert_equal [5, 4], recons.shape
+    assert_equal objects.to_a, recons.to_a
+
+    # Reconstruct subset from middle
+    recons = index.reconstruct_n(1, 3)
+    assert_equal [3, 4], recons.shape
+    assert_equal objects[1..3, true].to_a, recons.to_a
```
Comment (ankane) [inline test/index_test.rb:365]: This single assertion should be enough to test the happy path. Also, please test for n = 0 and n > number of objects in index.

File: test/index_test.rb [hunk 4]
```diff
@@ -319,6 +349,98 @@ def test_reconstruct
...
+    assert_equal objects[2, true].to_a, recons[1, true].to_a
+    assert_equal objects[4, true].to_a, recons[2, true].to_a
+
+    # Reconstruct in different order
+    keys = Numo::Int64.cast([4, 1, 3])
+    recons = index.reconstruct_batch(keys)
+    assert_equal [3, 4], recons.shape
+    assert_equal objects[4, true].to_a, recons[0, true].to_a
+    assert_equal objects[1, true].to_a, recons[1, true].to_a
+    assert_equal objects[3, true].to_a, recons[2, true].to_a
```
Comment (ankane) [inline test/index_test.rb:406]: This set of assertions should be enough to test the happy path. Please also test for empty ids and non-existent ids.

**Review by ankane — COMMENTED** [review]
Thanks @jan-kozinski. Added some comments inline.

**Comment by jan-kozinski** [comment]
thanks for the review I'll get back to this soon

File: test/index_test.rb (hunk 3 above)
Comment (jan-kozinski) [reply test/index_test.rb:365]: when called with `n > ntotal`  faiss raises runtime errors.  I think it would be nice to have more idiomatic ruby errors such as "IndexError", but this would either require duplicating faiss code and risking bugs or catching faiss errors and matching the error messages which also seems to add an unnecessary coupling on faiss internals, so I left the RuntimeError

**Comment by ankane** [comment]
Awesome, thanks @jan-kozinski! Merged a version of this in the commit above.

File: test/index_test.rb (hunk 3 above)
Comment (ankane) [reply test/index_test.rb:365]: It looks like `IndexBinaryFlat` indexes (and possibly others) don't have validation and will perform an out-of-bounds read, so added checks with an `IndexError` in b382867b4f1566d1940610eba6bb5cea89b6f60c.
